[SSO] Update Sign In form to enforce SSO (ONLY when flag in localsettings is enabled)

biyeun opened the pull request:
## Summary
Review commit-by-commit.
This addresses the following ticket: https://dimagi-dev.atlassian.net/browse/SS-101

The Sign In user experience has been altered to ensure that we enforce SSO login for users whose email domain is tied to an active Identity Provider.

NOTE: This is only applicable when the `ENFORCE_SSO_LOGIN` in `localsettings` is set to `True` (it is `False` by default). Changes were very carefully applied to ensure that the current login experience is not altered by this update. Once reviewed, this code is safe to merge into production without QA as production will not have `ENFORCE_SSO_LOGIN` enabled.

## Feature Flag
more like a `localsettings` "flag": `ENFORCE_SSO_LOGIN`

## Product Description
<img width="591" alt="Screen Shot 2021-05-12 at 2 41 35 PM" src="https://user-images.githubusercontent.com/716573/118193255-0405de00-b40d-11eb-9869-418437865e1b.png">
<img width="521" alt="Screen Shot 2021-05-12 at 2 41 52 PM" src="https://user-images.githubusercontent.com/716573/118193276-0b2cec00-b40d-11eb-8c52-d8d13602dbac.png">
<img width="523" alt="Screen Shot 2021-05-12 at 2 41 57 PM" src="https://user-images.githubusercontent.com/716573/118193296-1253fa00-b40d-11eb-8934-f3fb2f1c896f.png">
<img width="698" alt="Screen Shot 2021-05-12 at 2 46 50 PM" src="https://user-images.githubusercontent.com/716573/118193309-1718ae00-b40d-11eb-8207-597ea7821f34.png">


## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
In the process of assessing difficulty for writing JS tests for this. I am about to upgrade `sinon`, so I think I will dive into JS tests on a separate PR.

### QA Plan
QA will happen once all the SSO changes have been made and merged.

### Safety story
This feature is not active unless `ENFORCE_SSO_LOGIN` is set to `True` in `localsettings`, so it does not affect any existing active workflow.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

biyeun committed:
add a setting for enforcing SSO login / signup

biyeun committed:
pass enforce_sso_login bool to template context if applicable

Note: We only want to do this for the main authentication form and not on any two factor auth forms.

biyeun committed:
add shortcut method to get the IdP Login URL

biyeun committed:
accept username as a GET parameter

I will deal with removing the session stored username stuff at the end of development if no longer needed.

biyeun committed:
add method to get required identity provider for a username

biyeun committed:
add util view to check sso login requirements for a given username

biyeun committed:
add binding handler for calling the given accessor on enter key press

biyeun committed:
create knockout controller to manage login form when SSO login is enforced

biyeun committed:
just changing the indentation

biyeun committed:
prepare template for knockout controller

biyeun committed:
prepare form for knockout controller

biyeun committed:
initialize the login controller only when enforcing SSO is enabled

biyeun committed:
make sure SSO users can't login the old way if javascript fails

czue commented on corehq/apps/sso/models.py:
```diff
@@ -136,6 +137,19 @@ def get_sso_exempt_users(self):
+        """
+        Gets the login endpoint for the IdentityProvider based on the protocol
+        being used. Since we only support SAML2 right now, this redirects to
+        the SAML2 login endpoint.
```
Optional: did you consider just making this explicitly SAML2 until there's a need for multiple versions? this comment has a little bit of "yagni" feel to it (though I have no idea what our roadmap looks like for adding other options)

biyeun replied:
it's still a handy shortcut for the username so I don't think this is full "yagni", just a thoughtful comment. Also, we will be supporting other Identity Providers in the future, and it's unclear what their support will be

czue replied:
Got it. Also probably I could have been more clear in my comment: I just meant renaming the function to `get_saml2_login_url`.

czue commented on corehq/apps/registration/static/registration/js/user_login_form.js:
```diff
@@ -0,0 +1,147 @@
+        self.updateContinueText = function (expandPasswordField = false) {
+            self.continueTextPromise = $.post(self.checkSsoLoginStatusUrl, {
+                username: self.authUsername(),
+            }, function (data) {
```
out of curiosity was there a reason this was implemented as a post instead of passing the context to the initial template?

biyeun replied:
yes, we don't want to present the entire list of supported IdP's and email domains to be read in javascript for security/privacy purposes. This was discussed outside of this thread.

czue commented on corehq/apps/registration/static/registration/js/user_login_form.js:
```diff
@@ -0,0 +1,147 @@
+    'hqwebapp/js/assert_properties',
+    'hqwebapp/js/utils/email',
+    'hqwebapp/js/knockout_bindings.ko',
+], function (
```
is the long term plan that this replaces the current login form? or we'll maintain both indefinitely?

biyeun replied:
I think we need to have both because not all environments want to enforce SSO by default. If we decide that this is a non-issue, it's easy enough to remove the flag and we're on our way.

biyeun replied:
it doesn't replace but adds on to `registration/js/login`. I will likely need to use it for the login form when a session has timed out as well...so this isn't the only place it will be used.

czue reviewed (commented):
This looks great to me overall and I appreciated having everything clearly broken up by commit.

Just had a couple drive-by thoughts/questions. Main one is whether `ENFORCE_SSO_LOGIN` is meant to be a "permanent" setting, or it's more just a QA transition flag that will eventually be removed.

I held off on approval only because I don't think I know the front end code and this feature well enough to feel confident about catching any possible bugs/regressions so I am hoping someone more familiar will review. But everything looks great from my side.

millerdev commented on corehq/apps/hqwebapp/views.py:
```diff
@@ -1368,3 +1382,29 @@ def post(self, request, *args, **kwargs):
+    sso_url = None
+    continue_text = None
+
+    idp = IdentityProvider.get_required_identity_provider(username)
+    if idp:
+        is_sso_required = True
+        sso_url = idp.get_login_url(username=username)
+        continue_text = _("Continue to {}").format(idp.name)
```
Should this be evaluated if `ENFORCE_SSO_LOGIN` is false?

biyeun replied:
no, as this has nothing to do with forcing someone to login or not—that is all handled by the sign up or sign in views themselves. This is just an informational/utility view and should accurately report the status.

millerdev reviewed (approved):
There is some new Python code in this PR that does not have tests. Ex: a test could be added for `HqLoginView` to ensure it redirects when an SSO login is required and does not redirect under similar circumstances when `ENFORCE_SSO_LOGIN` is false.

biyeun committed:
feedback

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

biyeun commented on corehq/apps/registration/static/registration/js/user_login_form.js:
```diff
@@ -0,0 +1,147 @@
+                    if (expandPasswordField) self.continueToPasswordLogin();
+                }
+            })
+                .fail(function () {
+                    self.continueButtonText(self.defaultContinueText);
+                    if (expandPasswordField) self.continueToPasswordLogin();
+                });
```
how is this non-standard? it's related to the promise above, hence the indentation. I have seen this used / have used it all the time

millerdev replied:
I expected
```diff
-            })
-                .fail(function () {
+            }).fail(function () {
                 ...
```

biyeun committed:
feedback

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

biyeun committed:
Update corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

biyeun committed:
nit

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

biyeun commented:
@millerdev thank you! yeah i'm working on those python tests already, I just wanted to make sure everyone was on board with these changes :)

biyeun committed:
add tests for SSO enforcement in login views and forms

Also ensures that when ENFORCE_SSO_LOGIN = False, the current production behavior remains as expected.

biyeun commented:
alright, thank you @millerdev and @czue for your speedy feedback! Since everything looks good code-wise, I've added some tests to further ensure everything remains solid. ⚡

biyeun committed:
ding dong the `which` is dead
